src/sources/util/unix.rs

```diff
@@ -19,3 +33,171 @@ pub fn change_socket_permissions(path: &Path, perms: Option<u32>) -> crate::Resu
     }
     Ok(())
 }
+
+/// This is a structure which represents what kind of metadata should be
+/// _collected_ by unix_stream.rs & unix_datagram.rs. I would do this with
+/// some kind of flags structure, but Rust doesn't have one, so I guess a
+/// struct-of-booleans works.
+#[derive(Default, Copy, Clone, Debug)]
+pub struct UnixSocketMetadataCollectTypes {
+    /// Use getpeername(2) (on stream sockets) or read the struct sockaddr
+    /// argument from recvfrom(2) (on datagram sockets) to get the bound name
+    /// of the other half of the socket.
+    pub peer_path: bool,
+
+    /// Uses fstat(2) to get the inode/device of the connected socket (in
+    /// stream socket mode). Ignored in datagram socket mode (it would simply
+    /// return the same inode/dev every time, for the listener socket)
+    pub socket_inode: bool,
+
+    /// Use getsockopt(2) SO_PEERCRED or LOCAL_PEERCRED etc to query the creds
+    /// of the process that opened the socket
+    pub peer_creds: bool,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct SocketInode {
+    pub dev: u64,
+    pub ino: u64,
+}
+
+impl Into<Value> for SocketInode {
+    fn into(self) -> Value {
+        let mut map = BTreeMap::new();
+        // really want "as i64" here - actually despite claiming to return
+        // a u64, the value we get from fstat(2) for dev on some platforms
+        // is the bit pattern for -1. So, reinterpret the "u64" as signed.
+        map.insert("dev".to_string(), Value::Integer(self.dev as i64));
+        map.insert("ino".to_string(), Value::Integer(self.ino as i64));
+        Value::Object(map)
+    }
+}
```

Comment:
I'm not seeing a ton of value in having a struct + impl for this logic, when the `socket_inode` in `UnixSocketMetadata` could just be an `Option<Value>` , and the place where this is constructed could directly instantiate a Value with the data, and not have to do the map `insert()` operations.